### PR TITLE
refactor(main): wrap processor-dependent code into a switch struct

### PR DESCRIPTION
**Pull request by anarchuser**

### CODE REVIEW

File: thorn/src/kernel/kernel.c [hunk 1]
```diff
@@ -57,34 +57,38 @@ void kernel_process () {
...
-            if (!fb) {
-                printf ("Error: Invalid Framebuffer received\r\n");
-            } else {
-                printf ("Received framebuffer: %p\r\n", fb);
-            }
-        }
+        kernel_init ();
     }
 
     while (processor_id != current_processor)
```
Comment (AtomToast) [inline thorn/src/kernel/kernel.c:94]: Everything that isn't the current processor hangs?

File: thorn/src/kernel/kernel.c [hunk 2]
```diff
@@ -93,25 +97,27 @@ void kernel_main (int processor_id) {
...
-        //        if (res < 0) {
-        //            printf ("error while starting kernel process");
-        //            return;
-        //        }
-        //
-        //        while (1) {
-        //            schedule ();
-        //        }
-        LOG ("DONE PRINTING");
+    while (current_processor != 3)
```
Comment (AtomToast) [inline thorn/src/kernel/kernel.c:100]: also what's the issue with processor 3?

Comment (anarchuser) [reply thorn/src/kernel/kernel.c:100]: I don't know. Maybe we accidentally awaken processor 3 prematurely and it crashes

File: thorn/src/kernel/kernel.c [hunk 3]
```diff
@@ -93,25 +97,27 @@ void kernel_main (int processor_id) {
     printf ("Hello, from processor %d\n\r", processor_id);
 
     current_processor++;
```
[the remark this answers is not in the document]
Comment (anarchuser) [reply thorn/src/kernel/kernel.c:99]: as mentioned above, this is a simple form of synchronisation barrier. It ensures every non-zero core waits for finished initialisation and waits for the previous core to have finished printing

File: thorn/src/kernel/kernel.c (hunk 2 above)
Comment (AtomToast) [reply thorn/src/kernel/kernel.c:100]: But what about the other 2 non-main processors? At least right now this doesn't do anything anyways

File: thorn/src/kernel/kernel.c [hunk 4]
```diff
@@ -93,25 +97,27 @@ void kernel_main (int processor_id) {
...
+            }
+        }
+        case 1:
+            draw ();
+            break;
+        case 2:
+        case 3:
+        default:
+            printf ("Undefined behaviour on processor %d\r\n", processor_id);
+            while (1);
```
[the remark this answers is not in the document]
Comment (anarchuser) [reply thorn/src/kernel/kernel.c:120]: no. This code is run by every processor

File: thorn/src/kernel/kernel.c (hunk 1 above)
Comment (AtomToast) [reply thorn/src/kernel/kernel.c:94]: is a static variable actually shared between cores?

Comment (anarchuser) [reply thorn/src/kernel/kernel.c:94]: yes

Comment (anarchuser) [reply thorn/src/kernel/kernel.c:94]: a static variable is guaranteed to only ever be initialised once

Comment (AtomToast) [reply thorn/src/kernel/kernel.c:94]: got the code now :+1:

File: thorn/src/kernel/kernel.c (hunk 2 above)
Comment (anarchuser) [reply thorn/src/kernel/kernel.c:100]: me neither. I'm looking into it right now

**Review by AtomToast — DISMISSED** [review]

**Pull request merged**